### PR TITLE
Update scrape_walmart.py

### PR DESCRIPTION
**Pull request by ryandomalley**
Started a price scraper for Walmart.com
Pulls the title, price, and rating from the top 10 results of a query

### CODE REVIEW

**Comment by firejake308** [comment]
@ryandomalley Do you know how to use `git rebase`? If so, could you rebase this onto the `linters` branch? There are a couple of code style issues here that the linter would highlight pretty easily. Also, if you use VSCode, you can setup Pylint to run automatically, so I can post instructions for that if you want as well.

If you've never `git rebase`d before and need help, just DM me

File: backend/scraper/scrape_walmart.py [hunk 1]
```diff
@@ -1 +1,88 @@
...
+        information = {
+            'product_title': product_title,
+            'product_identifier': 'requests-version',
+            'price': price,
+            'ratings': rating,
+            'stock_status': 'requests-version'
+        }
+        return information
+
+    def destructor(self):
```
Comment (firejake308) [inline backend/scraper/scrape_walmart.py:76]: Aren't Python destructors defined with `__del__()`?

File: backend/scraper/scrape_walmart.py [hunk 2]
```diff
@@ -1 +1,88 @@
...
+    def getProductUrls(self, queryString):
+        self.driver.get('https://www.walmart.com/search/?query=' + queryString)
+        URLS = []
+        for link in self.driver.find_element_by_class_name('search-product-result').find_elements_by_tag_name('a'):
+            if (link.get_attribute('class') == 'search-result-productimage gridview display-block'):
+                URLS.append(link.get_attribute('href'))
+        return URLS
+
+    #Requests + selenium combo?
+    def parseProductLink(self, url, session):
```
Comment (firejake308) [inline backend/scraper/scrape_walmart.py:44]: Let's use snake case names for functions

File: backend/scraper/scrape_walmart.py [hunk 3]
```diff
@@ -1 +1,88 @@
...
+from bs4 import BeautifulSoup
+import requests
+
+from selenium.webdriver.chrome.options import Options
+from selenium.webdriver.common.keys import Keys
+import time
+
+ZIPCODE = '77841'
+
+driver_path = r'C:/Users/Ryan/Desktop/Web Drivers/chromedriver.exe'
```
Comment (firejake308) [inline backend/scraper/scrape_walmart.py:11]: There's probably a better way to handle this than hard-coding the path. Usually we use environment variables or config files

File: backend/scraper/scrape_walmart.py [hunk 4]
```diff
@@ -1 +1,88 @@
...
+        page = session.get(url, headers={"User-Agent": "Mozilla/5.0"})
+        soup = BeautifulSoup(page.text, 'lxml')
+        try:
+            product_title = soup.find(class_='prod-ProductTitle font-normal').get_text()
+        except:
+            product_title = 'ERROR: ' + url
+        try:
+            rating = soup.find(class_='ReviewsHeader-ratingPrefix font-bold').get_text()
+        except:
+            rating = 'ERROR: ' + url
```
Comment (firejake308) [inline backend/scraper/scrape_walmart.py:55]: This is fine for development, but in production, I wouldn't want users to see the URL where the rating is supposed to be. Instead, just `print('ERROR: ' + url)` and set rating to `None`

File: backend/scraper/scrape_walmart.py (hunk 2 above)
Comment (ryandomalley) [reply backend/scraper/scrape_walmart.py:44]: Snake case for both functions and variables, or just functions?

Comment (firejake308) [reply backend/scraper/scrape_walmart.py:44]: Let's do both functions and variables. PascalCase should be reserved for classes

**Comment by firejake308** [comment]
@rynbhuiya I don't think you're tracking the right files. Your latest commit didn't actually modify any of the Python/code files, but it did add an `.exe` and a `.log`, which I would actually prefer not to track. Since each of the members uses a different browser/operating system, it would be better for everyone to just install their own drivers.

**Comment by firejake308** [comment]
@ryandomalley @rynbhuiya Besides the driver that I can download from [here](https://www.selenium.dev/documentation/en/webdriver/driver_requirements/), what else do I need to download to use Selenium?

**Comment by firejake308** [comment]
image
Okay, so I got this working on my computer more or less, but I get this error. Is `lxml` a library we need to `pip install`? If so, we should add it to `requirements.txt`

**Comment by firejake308** [comment]
```
Item Name: Sweet Apple Slices, 32 oz
Product ID: requests-version
Price: 5.78
Rating: 3.1
Stock Status: requests-version
```
Also getting these logs. Price and item name look fine, but it looks like Product ID and Stock Status aren't working

File: backend/scraper/scrape_walmart.py [hunk 5]
```diff
@@ -1 +1,100 @@
...
+
+import sys
+
+class WalmartScraper:
+    zipSet = False
+    def __init__(self):
+        chrome_options = Options()
+        chrome_options.add_argument('--headless')
+        try:
+            self.driver = webdriver.Firefox(executable_path=r'.\geckodriver.exe')
```
Comment (firejake308) [inline backend/scraper/scrape_walmart.py:17]: Can we get it so that this script takes in an argument of either `chrome` or `firefox` that lets it choose which browser to use?

File: backend/scraper/scrape_walmart.py [hunk 6]
```diff
@@ -1 +1,100 @@
...
+        for link in self.driver.find_element_by_class_name('search-product-result').find_elements_by_tag_name('a'):
+            if (link.get_attribute('class') == 'search-result-productimage gridview display-block'):
+                URLS.append(link.get_attribute('href'))
+        return URLS
+
+    #Requests + selenium combo?
+    def parse_product_link(self, url, session):
+        print('Opening url: ' + url) #debugging
+        page = session.get(url, headers={"User-Agent": "Mozilla/5.0"})
+        soup = BeautifulSoup(page.text, 'lxml')
```
Comment (firejake308) [inline backend/scraper/scrape_walmart.py:48]: I think we need to add `lxml` to the `requirements.txt`

File: backend/scraper/scrape_walmart.py [hunk 7]
```diff
@@ -1 +1,100 @@
...
+        try:
+            self.driver.close()
+        except:
+            print('Driver does not exist')
+
+scraper = WalmartScraper()
+apples = scraper.get_product('apples', '77840')
+for item in apples:
+    print("Item Name: " + item['product_title'])
+    print("Product ID: " + item['product_identifier'])
```
Comment (firejake308) [inline backend/scraper/scrape_walmart.py:96]: As I mentioned in an earlier comment, Product ID and Stock Status don't seem to be working. I think we should either fix them or just leave them out.

File: backend/scraper/scrape_walmart.py [hunk 8]
```diff
@@ -1 +1,100 @@
...
+        }
+        return information
+
+    def __del__(self):
+        try:
+            self.driver.close()
+        except:
+            print('Driver does not exist')
+
+scraper = WalmartScraper()
```
Comment (firejake308) [inline backend/scraper/scrape_walmart.py:92]: This bit should be wrapped in an `if __name__ == "__main__":` block. Otherwise, it will be run every time `WalmartScraper` is imported from another file.

File: backend/scraper/scrape_walmart.py [hunk 9]
```diff
@@ -11,12 +11,25 @@
...
-            sys.exit("Web Driver is missing")
+        if len(sys.argv) != 2:
+            raise ValueError('Please enter a browser driver to use (Chrome or Firefox)')
+
+        browser = sys.argv[1]
+        if browser.lower() == 'chrome':
+            chrome_options = Options()
+            chrome_options.add_argument('--headless')
+            try:
+                driver = r'./chromedriver.exe'
```
Comment (firejake308) [inline backend/scraper/scrape_walmart.py:22]: Do you think you could use the WebDriver Manager, like Vicente did in #33? This PR has been open for a while, and using the manager isn't necessary for functionality, so we can do it in a separate PR, but eventually, it would be easier to use if the scripts manage their own drivers

File: backend/scraper/scrape_walmart.py [hunk 10]
```diff
@@ -49,7 +62,7 @@ def parse_product_link(self, url, session):
 
         # Product Title
         try:
-            product_title = soup.find(class_='prod-ProductTitle font-normal').get_text()
+            product_title = soup.find(class_='prod-ProductTitle prod-productTitle-buyBox font-bold').get_text()
```
Comment (firejake308) [inline backend/scraper/scrape_walmart.py:65]: Why was this change necessary? Has Walmart changed the layout, or was it wrong to begin with?

Comment (rynbhuiya) [reply backend/scraper/scrape_walmart.py:65]: It was working up until now, so I checked the web page and the class name was changed.

File: backend/scraper/scrape_walmart.py (hunk 9 above)
Comment (rynbhuiya) [reply backend/scraper/scrape_walmart.py:22]: Yes, I can take a look and see if I can make it work.
